amazon/ionbenchmark/benchmark_runner.py

```diff
@@ -112,23 +116,90 @@ def test_fn():
+            else:
+                def test_fn():
+                    with open(data_file, 'br') as f:
+                        for v in loader_dumper.load(f, parse_eagerly=False):
+                            pass
+        elif _format.format_is_json(format_option):
+            if return_obj:
+                def test_fn():
+                    rtn = []
+                    with open(data_file, 'r') as f:
+                        for v in loader_dumper.load(f):
+                            rtn.append(v)
+                    return rtn
+            else:
+                def test_fn():
+                    with open(data_file, 'r') as f:
+                        for v in loader_dumper.load(f):
+                            pass
+        elif _format.format_is_cbor(format_option):
+            if return_obj:
+                def test_fn():
+                    rtn = []
+                    with open(data_file, 'br') as f:
+                        for v in loader_dumper.load(f):
+                            rtn.append(v)
+                    return rtn
+            else:
+                def test_fn():
+                    with open(data_file, 'br') as f:
+                        for v in loader_dumper.load(f):
+                            pass
+        elif _format.format_is_protobuf(format_option):
+            # Refer to https://github.com/amazon-ion/ion-python/issues/326
+            raise NotImplementedError("Benchmarking Protocol Buffer multiple top level object use case may not "
+                                      "support yet.")
 
     elif match_arg == ['file', 'write', 'load_dump']:
+        # This method should return a list that holds all top_level values.
         data_obj = benchmark_spec.get_data_object()
         data_format = benchmark_spec.get_format()
-        if _format.format_is_binary(data_format) or _format.format_is_ion(data_format):
-            def test_fn():
-                with tempfile.TemporaryFile(mode="wb") as f:
-                    return loader_dumper.dump(data_obj, f)
+        if _format.format_is_protobuf(data_format):
+            # Refer to https://github.com/amazon-ion/ion-python/issues/326
+            raise NotImplementedError("Benchmarking Protocol Buffer multiple top level object use case may not "
+                                      "support yet.")
+        elif _format.format_is_cbor(data_format):
+            if custom_file:
+                def test_fn():
+                    with open(custom_file, 'ab') as f:
+                        loader_dumper.dump(data_obj, f)
+            else:
+                def test_fn():
+                    with tempfile.TemporaryFile(mode="ab") as f:
+                        loader_dumper.dump(data_obj, f)
+        elif _format.format_is_ion(data_format):
+            if custom_file:
+                def test_fn():
+                    with open(custom_file, 'ab') as f:
+                        loader_dumper.dump(data_obj, f, sequence_as_stream=True)
```

Comment:
The purpose of the `loader_dumper` abstraction is to encapsulate format stuff like `sequence_as_stream`. Why does that need to be passed here?

Reply:
Changed